### PR TITLE
Cabal doctest

### PR DESCRIPTION
**Pull request by phadej**
cc @RyanGlScott

### CODE REVIEW

File: bytes.cabal [hunk 1]
```diff
@@ -40,10 +40,9 @@ flag test-doctests
 custom-setup
   -- any should do
   setup-depends:
-    Cabal     >= 1.14,
-    base      >= 4.5     && <5,
-    directory >= 1.1.0.1,
-    filepath  >= 1.3.0.0
+    base          >= 4.5 && <5,
+    Cabal         >= 1.14, 
+    cabal-doctest >= 1 && <1.1
```
Comment (RyanGlScott) [inline bytes.cabal]: How does the `test-doctests` flag interact with this feature? That is, if someone configured this project with `-f-test-doctests`, would `custom-setup` also be disabled? (I don't think it's a dealbreaker if we can't make it do that, but I am curious.)

File: Setup.lhs [hunk 2]
```diff
@@ -160,5 +166,13 @@ generateBuildModule flags pkg lbi = do
...
 testDeps :: ComponentLocalBuildInfo -> ComponentLocalBuildInfo -> [(InstalledPackageId, PackageId)]
 testDeps xs ys = nub $ componentPackageDeps xs ++ componentPackageDeps ys
+#endif
+
+main :: IO ()
+main = defaultMainWithHooks simpleUserHooks
+  { buildHook = \pkg lbi hooks flags -> do
+     generateBuildModule "doctests" flags pkg lbi
+     buildHook simpleUserHooks pkg lbi hooks flags
+  }
```
Comment (RyanGlScott) [inline Setup.lhs]: This is a little wordier than I would have expected. Couldn't we export something like this from `cabal-doctest`:

```haskell
doctestsUserHooks :: UserHooks
doctestsUserHooks testsuiteName = simpleUserHooks
  { buildHook = \pkg lbi hooks flags -> do
     generateBuildModule "doctests" flags pkg lbi
     buildHook simpleUserHooks pkg lbi hooks flags
  }
```

And then just do `main = defaultMainWithHooks $ doctestUserHooks "doctests"`? There's far less boilerplate involved on the commonly taken path (modern `Cabal`) this way, and you can still override the fields of `doctestUserHooks "doctests"` if you wish.

Comment (phadej) [inline Setup.lhs]: Yeah, just thought about this omw home. Maybe even also `defaultMainWithDoctests`

File: bytes.cabal (hunk 1 above)
Comment (phadej) [inline bytes.cabal]: IIRC it was a design decision that `custom-setup` section doesn't support flags / conditionals. But I'm not sure. I'll check.

Comment (phadej) [inline bytes.cabal]: https://github.com/haskell/cabal/issues/4286

**Comment by phadej** [comment]
@RyanGlScott could you take another look. If it's ok, I'll release `cabal-doctest` to Hackage.

**Comment by RyanGlScott** [comment]
@phadej, looks great! Please go ahead and upload it to Hackage.

**Pull request merged**